mmseg/models/losses/focal_loss.py

```diff
@@ -0,0 +1,327 @@
+    target = target.type_as(pred)
+    one_minus_pt = (1 - pred_sigmoid) * target + pred_sigmoid * (1 - target)
+    focal_weight = (alpha * target + (1 - alpha) *
+                    (1 - target)) * one_minus_pt.pow(gamma)
+
+    loss = F.binary_cross_entropy_with_logits(
+        pred, target, reduction='none') * focal_weight
+    final_weight = torch.ones(1, pred.size(1)).type_as(loss)
+    if weight is not None:
+        if weight.shape != loss.shape and weight.size(0) == loss.size(0):
+            # For most cases, weight is of shape (N, ),
+            # which means it does not have the second axis num_class
+            weight = weight.view(-1, 1)
+        assert weight.dim() == loss.dim()
+        final_weight = final_weight * weight
+    if class_weight is not None:
+        final_weight = final_weight * class_weight
+    if valid_mask is not None:
+        final_weight = final_weight * valid_mask
+    loss = weight_reduce_loss(loss, final_weight, reduction, avg_factor)
+    return loss
+
+
+def sigmoid_focal_loss(pred,
+                       target,
+                       one_hot_target,
+                       weight=None,
+                       gamma=2.0,
+                       alpha=0.5,
+                       class_weight=None,
+                       valid_mask=None,
+                       reduction='mean',
+                       avg_factor=None):
+    r"""A warpper of cuda version `Focal Loss
+    <https://arxiv.org/abs/1708.02002>`_.
+    Args:
+        pred (torch.Tensor): The prediction with shape (N, C), C is the number
+            of classes.
+        target (torch.Tensor): The learning label of the prediction. It's shape
+            should be (N, )
+        one_hot_target (torch.Tensor): The learning label with shape (N, C)
+        weight (torch.Tensor, optional): Sample-wise loss weight.
+        gamma (float, optional): The gamma for calculating the modulating
+            factor. Defaults to 2.0.
+        alpha (float, torch.Tensor, optional): A balanced form for Focal Loss.
+            Defaults to 0.5.
+        class_weight (torch.Tensor, optional): Weight of each class.
+            Defaults to None.
+        valid_mask (torch.Tensor ,None): A mask uses 1 to mark the valid
+            samples and uses 0 to mark the ignored samples. Default: None.
+        reduction (str, optional): The method used to reduce the loss into
+            a scalar. Defaults to 'mean'. Options are "none", "mean" and "sum".
+        avg_factor (int, optional): Average factor that is used to average
+            the loss. Defaults to None.
+    """
+    # Function.apply does not accept keyword arguments, so the decorator
+    # "weighted_loss" is not applicable
+
+    # _sigmoid_focal_loss doesn't accept alpha of list type. Therefore, if
+    # a list is given, we set float_alpha as 0.5. This means setting equal
+    # weight for foreground class and background class.
+    float_alpha = alpha if isinstance(alpha, float) else 0.5
```

Comment:
```suggestion
   assert isinstance(alpha, float)
```
This way may make user input float explicitly, avoid ambiguity

Reply:
This won't implicitly change the input alpha of type list to float 0.5. In order to use the cuda version focal loss. I have to set alpha as 0.5. Then in the post-processing process, I offset the effect of setting alpha as 0.5 and use the user-provided alpha(a list) to regulate the loss.(line 118-121) Therefore, providing an alpha of list type is acceptable.

Reply:
What's more, this function is only called by class FocalLoss, who makes sure that only alpha of type float and tensor will be input to the function.

Reply:
I see. We may explain it in the comments. And maybe rename `float_alpha` to `pseudo_alpha`?

Reply:
ok